design-proposals/kubevirtbmc.md

```diff
@@ -0,0 +1,156 @@
+
+## Motivation
+
+Developers require the ability to deploy applications or systems in local virtual environments like bare-metal ones. Existing solutions involve libvirt domains or QEMU VMs with BMC emulators, which are not directly compatible with KubeVirt due to its internal use of [`virtqemud`](https://libvirt.org/manpages/virtqemud.html), necessitating a Kubernetes-native solution. This was initially raised in the GitHub issue: [Expose a virtual BMC service per virtual machine, for IPMI and/or Redfish actions · Issue #9108 · kubevirt/kubevirt](https://github.com/kubevirt/kubevirt/issues/9108).
+
+## Goals
+
+- Provide essential BMC functionalities such as (rebooting, changing boot devices, and mounting virtual media) via IPMI and Redfish protocols against VMs backed by KubeVirt
+- Expose service endpoints for clients to access either in-cluster or externally
+- Enable deployment tools, e.g. [Metal3](https://metal3.io/), [Tinkerbell](https://tinkerbell.org/), [Bifrost](https://opendev.org/openstack/bifrost/), etc., to install software on VMs without adapting to the virtual environment by modifying their source code
+
+## Non Goals
+
+- Providing BMC functionalities for bare-metal machines
+
+## Definition of Users
+
+- Users who want to do out-of-band management to the VMs within a KubeVirt environment directly using bare-metal-like methods, e.g., IPMI or Redfish
+- Users who want to install software like OpenShift or OpenStack on VMs locally within a KubeVirt environment using existing deployment tools
+
+## User Stories
+
+- As a developer, I want to deploy a multi-node OpenShift cluster using its native installer in KubeVirt VMs.
+- As a tester, I need to quickly spin up and destroy OpenStack clusters in a virtual environment for testing purposes.
+
+## Repos
+
+No existing repos under the KubeVirt organization will be impacted. KubeVirtBMC, proposed to be a new project under the KubeVirt organization, augments the KubeVirt ecosystem as an optional add-on providing an operator and the emulated BMC implementation.
+
+# Design
+
+KubeVirtBMC will act as a bridge between KubeVirt VMs and the BMC functionalities, using Kubernetes services to expose these capabilities. KubeVirtBMC consists of two components:
+
+- **vmbmc-controller**: A typical Kubernetes controller reconciles the VirtualMachineBMC, VirtualMachine, and Service objects
+- **vmbmc**: A BMC simulator for serving IPMI and Redfish endpoints and translating the incoming requests to native KubeVirt API calls
```

Comment:
Other controllers / Pods in KubeVirt have a named prefixed with `virt-`. Maybe it makes sense to prefix these names too? E.g. `virt-bmc-controller` and `virt-bmc`?

Reply:
Wouldn't that suggest they were somehow core to KubeVirt and maybe managed by virt-operator?

Reply:
True, but that argument could apply to the CRD names too?

Reply:
The CRD has now been renamed `VirtualMachineBMC`. `KubeVirtBMC` is the project name. I'm okay with renaming the components if that follows the convention of the KubeVirt ecosystem.